ts/components/messages/paginated/MessageList.tsx

```diff
@@ -0,0 +1,285 @@
+import { EdgeBorderComponent } from "../../screens/EdgeBorderComponent";
+import { UIMessage } from "../../../store/reducers/entities/messages/types";
+
+import MessageListItem from "./MessageListItem";
+
+type ItemLayout = {
+  length: number;
+  offset: number;
+  index: number;
+};
+
+export type AnimatedProps = {
+  animated?: {
+    onScroll: (_: NativeSyntheticEvent<NativeScrollEvent>) => void;
+    scrollEventThrottle?: number;
+  };
+};
+
+type OwnProps = {
+  messages: ReadonlyArray<UIMessage>;
+  refreshing: boolean;
+  onRefresh: () => void;
+  onLoadMore?: () => void;
+  onPressItem: (id: string) => void;
+  onLongPressItem: (id: string) => void;
+  selectedMessageIds: Option<Set<string>>;
+  ListEmptyComponent?: React.ComponentProps<
+    typeof FlatList
+  >["ListEmptyComponent"];
+};
+
+type Props = OwnProps & AnimatedProps;
+
+type State = {
+  prevMessages?: OwnProps["messages"];
+  itemLayouts: ReadonlyArray<ItemLayout>;
+  longPressedItemIndex: Option<number>;
+  isFirstLoad: boolean;
+};
+
+const ITEM_WITHOUT_CTABAR_HEIGHT = 114;
+const ITEM_LOADING_HEIGHT = ITEM_WITHOUT_CTABAR_HEIGHT;
+const ITEM_SEPARATOR_HEIGHT = 1;
+
+const styles = StyleSheet.create({
+  itemLoadingContainer: {
+    height: ITEM_LOADING_HEIGHT,
+    paddingVertical: 16,
+    paddingHorizontal: customVariables.contentPadding,
+    flex: 1
+  },
+  itemLoadingHeaderWrapper: {
+    flexDirection: "row",
+    marginBottom: 4
+  },
+  itemLoadingHeaderCenter: {
+    flex: 1,
+    paddingRight: 55 // Includes right header space
+  },
+  itemLoadingContentWrapper: {
+    flexDirection: "row",
+    alignItems: "center",
+    height: 42
+  },
+  itemLoadingContentCenter: {
+    flex: 1,
+    paddingRight: 32
+  },
+  padded: {
+    paddingHorizontal: customVariables.contentPadding
+  },
+  activityIndicator: {
+    padding: 12
+  }
+});
+
+const generateItemLayouts = (
+  messages: OwnProps["messages"]
+): Array<ItemLayout> => {
+  // eslint-disable-next-line
+  let offset = 0;
+  // eslint-disable-next-line
+  const itemLayouts: ItemLayout[] = [];
+  // eslint-disable-next-line functional/no-let
+  let itemHeightWithSeparator =
+    ITEM_WITHOUT_CTABAR_HEIGHT + ITEM_SEPARATOR_HEIGHT;
+
+  for (
+    // eslint-disable-next-line functional/no-let
+    let i = 0;
+    i < messages.length;
+    i += 1, offset += itemHeightWithSeparator
+  ) {
+    itemHeightWithSeparator =
+      i === messages.length - 1
+        ? ITEM_WITHOUT_CTABAR_HEIGHT
+        : ITEM_WITHOUT_CTABAR_HEIGHT + ITEM_SEPARATOR_HEIGHT;
+
+    // eslint-disable-next-line functional/immutable-data
+    itemLayouts.push({
+      length: itemHeightWithSeparator,
+      offset,
+      index: i
+    });
+  }
+
+  return itemLayouts;
+};
+
+const ItemSeparator = () => <ItemSeparatorComponent noPadded={true} />;
+
+const AnimatedFlatList = Animated.createAnimatedComponent(FlatList);
+
+class MessageList extends React.Component<Props, State> {
```

Comment:
I eventually re-wrote most of this not just to reflect our style (prefer FC over Classes) but also because addressing some issues and writing tests led to such a large refactoring.